### PR TITLE
dockerize carla + openpilot

### LINKED ISSUE #1349: Communication Error on sim with carla
**Describe the bug**
Followed the instructions to get openpilot running with Carla. Once the bridge starts, opendbc seems to have issues with the counter of the signals:
```bash
...
0x17C COUNTER FAIL 2 -- 3 vs 3
0x17C COUNTER FAIL 2 -- 0 vs 0
0x17C COUNTER FAIL 2 -- 1 vs 1
0x17C COUNTER FAIL 2 -- 2 vs 2
...
```
As far as I can tell, the signal counter does not increase. Yet, when creating the can messages the idx does increase.

Also os.read() seems to have issues reading from stdin as the input size to input does not have the proper input size (but I assume this is a follow up problem of the one above?):
```
Traceback (most recent call last):
  File "/home/jodo/openpilot/selfdrive/modeld/runners/keras_runner.py", line 57, in <module>
    run_loop(m)
  File "/home/jodo/openpilot/selfdrive/modeld/runners/keras_runner.py", line 32, in run_loop
    ret = m.predict_on_batch(idata)
...
ValueError: Cannot reshape a tensor with 0 elements to shape [1,12,128,256] (393216 elements) for 'model/reshape/Reshape' (op: 'Reshape') with input shapes: [1,0], [4] and with input tensors computed as partial shapes: input[1] = [1,12,128,256].
```

**How to reproduce or log data**
Run openpilot with Carla as described [here](https://github.com/commaai/openpilot/tree/master/tools/sim).

**Device/Version information (please complete the following information):**
 - Device: Linux Mint 19.2
 - Version: latest master (85a09b3113fe45b5feee7c1c421e2c43cb2ac03a)
 - Car make/model: carla sim

### LINKED ISSUE #1655: modeld crashes when running Carla simulation
**Describe the bug**
modeld crashes expecting non-empty input from stdin when running Carla simulation.

**How to reproduce or log data**
Follow steps from [tools/sim/README](https://github.com/commaai/openpilot/blob/master/tools/sim/README.md)

**Expected behavior**
Openpilot works with Carla

**Device/Version information (please complete the following information):**
Ubuntu 18.04

**Additional context**
keras_runner receives an empty input from stdin and fails:

Traceback (most recent call last):                                                                 
  File "/home/dake/openpilot/selfdrive/modeld/runners/keras_runner.py", line 49, in <module> 
    run_loop(m)
  File "/home/dake/openpilot/selfdrive/modeld/runners/keras_runner.py", line 32, in run_loop
    inputs.append(read(ts).reshape(shp))
  File "/home/dake/openpilot/selfdrive/modeld/runners/keras_runner.py", line 16, in read
    assert(len(st) > 0)
AssertionError

### LINKED ISSUE #1672: Errors when running with Carla in a docker container
**Describe the bug**

I'm trying to run the openpilot simulation but several errors appear in the manager when I start the carla bridge.

Here's the full manager.py output: https://pastebin.com/raw/T0Rq0hAZ
When I start the bridge, the UI window disappears.
Three of the threads are marked in red at first: logmessaged, ui, modeld
Then four of them are red: logmessaged, ui, modeld, proclogd

Here are some of the error messages:
```
CalibrationParams file found but error encountered
Traceback (most recent call last):
  File "/openpilot/selfdrive/locationd/calibrationd.py", line 71, in __init__
    self.vp = vp_from_rpy(calibration_params["calib_radians"])
KeyError: 'calib_radians'
```

```
chrt: failed to set pid 452's policy: Operation not permitted
```

```
./_modeld: error while loading shared libraries: libSNPE.so: cannot open shared object file: No such file or directory
```
(That comment about the libSNPE error may be related: https://github.com/commaai/openpilot/issues/1364#issuecomment-639340702)

Maybe the scons code only removes the SNPE model and library if the architecture is `Darwin`.

**How to reproduce or log data**

Here is the Dockerfile for the container: https://gist.github.com/FHof/2c06f9b353f2caa37a448adcf0d70607
I created the docker image with a recent openpilot version: commit c6bd1874503b26e8443ecb28afd6819b44684e78

**Expected behavior**
I assume all threads should be marked green and the window should not close.

### PR DESCRIPTION
**Pull request by nuwandavek**
**Description** : Dockerize openpilot to run on Dockerized Carla

- Dockerfile.openpilot : nvidia/opengl can be stripped down
- Hacky changes in the following files : selfdrive/camerad/main.cc,  selfdrive/controls/controlsd.py, selfdrive/manager.py, selfdrive/modeld/modeld, tools/sim/bridge.py, 

- Other changes : tools/sim/lib/keyboard_ctrl.py, tools/sim/start_carla_docker.sh

### CODE REVIEW

**Comment by adeebshihadeh** [comment]
Let's make sure we address all the issues with the simulation tag and any relevant PC tagged issues.

File: selfdrive/controls/controlsd.py [hunk 1]
```diff
@@ -207,7 +207,7 @@ def update_events(self, CS):
     if not self.sm['pathPlan'].mpcSolutionValid:
       self.events.add(EventName.plannerError)
     if not self.sm['liveLocationKalman'].sensorsOK and os.getenv("NOSENSOR") is None:
-      if self.sm.frame > 5 / DT_CTRL:  # Give locationd some time to receive all the inputs
+      if self.sm.frame > 5 / DT_CTRL:  # Give locationd som`2`1e time to receive all the inputs
```
Comment (pd0wm) [inline selfdrive/controls/controlsd.py:210]: Remove this change

File: selfdrive/controls/controlsd.py [hunk 2]
```diff
@@ -230,11 +230,12 @@ def update_events(self, CS):
     if self.sm['plan'].fcw:
       self.events.add(EventName.fcw)
     if self.sm['model'].frameDropPerc > 1:
-      self.events.add(EventName.modeldLagging)
+      #self.events.add(EventName.modeldLagging)
```
Comment (pd0wm) [inline selfdrive/controls/controlsd.py:233]: This needs to be gated between an environment variable. Maybe `SIMULATION=1`

File: selfdrive/controls/controlsd.py [hunk 3]
```diff
@@ -230,11 +230,12 @@ def update_events(self, CS):
...
+      pass
 
     # Only allow engagement with brake pressed when stopped behind another stopped car
     if CS.brakePressed and self.sm['plan'].vTargetFuture >= STARTING_TARGET_SPEED \
-       and not self.CP.radarOffCan and CS.vEgo < 0.3:
+and not self.CP.radarOffCan and CS.vEgo < 0.3:
```
Comment (pd0wm) [inline selfdrive/controls/controlsd.py:238]: don't change this whitespace

File: tools/sim/Dockerfile.sim [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    ;
+
+# Install Intel OpenCL drivers
+# Based on the Intel OpenCL installation instructions and the Intel docker file
+# in https://github.com/chihchun/opencl-docker
+# This seems to work for the CPU only.  
+ARG INTEL_DRIVER=opencl_runtime_16.1.1_x64_ubuntu_6.4.0.25.tgz
+ARG INTEL_DRIVER_URL=http://registrationcenter-download.intel.com/akdlm/irc_nas/9019
+
+#ADD http://registrationcenter-download.intel.com/akdlm/irc_nas/11396/SRB5.0_linux64.zip /tmp/intel-opencl-driver
```
Comment (pd0wm) [inline tools/sim/Dockerfile.sim:25]: Remove unused lines in dockerfile.

File: tools/sim/bridge.py [hunk 5]
```diff
@@ -64,60 +83,81 @@ def health_function():
...
     pm.send('health', dat)
-    rk.keep_time()
+    time.sleep(0.5)
+
+def fake_gps():
+  # TODO: read GPS from CARLA
+  pm = messaging.PubMaster(['gpsLocationExternal'])
+  while 1:
+    dat = messaging.new_message('gpsLocationExternal')
+    if dat is not None:
```
Comment (pd0wm) [inline tools/sim/bridge.py:93]: This should never be none

File: tools/sim/bridge.py [hunk 6]
```diff
@@ -64,60 +83,81 @@ def health_function():
...
+  pm = messaging.PubMaster(['driverState','dMonitoringState'])
   while 1:
+
+    # dmonitoringmodeld output
     dat = messaging.new_message('driverState')
     dat.driverState.faceProb = 1.0
     pm.send('driverState', dat)
-    time.sleep(0.1)
+
+    # dmonotirongd output
```
Comment (pd0wm) [inline tools/sim/bridge.py:107]: typo

File: tools/sim/bridge.py [hunk 7]
```diff
@@ -64,60 +83,81 @@ def health_function():
...
 
+  max_steer_angle = vehicle.get_physics_control().wheels[0].max_steer_angle
+
   # make tires less slippery
-  wheel_control = carla.WheelPhysicsControl(tire_friction=5)
+  # wheel_control = carla.WheelPhysicsControl(tire_friction=5)
   physics_control = vehicle.get_physics_control()
-  physics_control.mass = 1326
-  physics_control.wheels = [wheel_control]*4
+  physics_control.mass = 126
```
Comment (pd0wm) [inline tools/sim/bridge.py:155]: Change mass to something reasonable

File: tools/sim/start_carla.sh [hunk 8]
```diff
@@ -1,18 +1,20 @@
...
-if [ ! -d carla ]; then
-  rm -rf carla_tmp
-  mkdir -p carla_tmp
-  cd carla_tmp
-  tar xvf ../$FILE
-  easy_install PythonAPI/carla/dist/carla-0.9.7-py3.5-linux-x86_64.egg || true
-  cd ../
-  mv carla_tmp carla
-fi
+# Requires nvidia docker - https://github.com/NVIDIA/nvidia-docker
```
Comment (pd0wm) [inline tools/sim/start_carla.sh:3]: Find a good check if `nvidia-docker` is installed. Maybe something with `apt`?

Comment (adeebshihadeh) [reply tools/sim/start_carla.sh:3]: It’s commented out below this line

File: selfdrive/controls/controlsd.py (hunk 2 above)
Comment (adeebshihadeh) [reply selfdrive/controls/controlsd.py:233]: hmm, this should run fast enough with the keras runner

File: tools/sim/Dockerfile.sim [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+COPY ./rednose $HOME/openpilot/rednose
+COPY ./common $HOME/openpilot/common
+COPY ./models $HOME/openpilot/models
+COPY ./opendbc $HOME/openpilot/opendbc
+COPY ./cereal $HOME/openpilot/cereal
+COPY ./panda $HOME/openpilot/panda
+COPY ./selfdrive $HOME/openpilot/selfdrive
+COPY ./tools $HOME/openpilot/tools
+
+RUN apt-get update && apt-get install -y beignet-opencl-icd
```
Comment (adeebshihadeh) [inline tools/sim/Dockerfile.sim:122]: move this above the copies

File: tools/sim/start_carla.sh (hunk 8 above)
Comment (pd0wm) [reply tools/sim/start_carla.sh:3]: yeah, the `which nvidia-docker` check doesn't work, that's why it's commented out.

**Pull request merged**